dor/adapters/event_store.py

```diff
@@ -0,0 +1,84 @@
+import uuid
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Optional
+
+from sqlalchemy import DateTime, String, select, Uuid
+from sqlalchemy.orm import Mapped
+from sqlalchemy.orm import mapped_column
+
+from dor.domain import models
+from dor.adapters.catalog import Base
+
+
+class EventStore(ABC):
+
+    @abstractmethod
+    def add(self, event: models.WorkflowEvent):
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_all_by_tracking_identifier(self, tracking_identifier: str) -> list[models.WorkflowEvent]:
+        raise NotImplementedError
+
+
+class MemoryEventStore(EventStore):
+
+    def __init__(self):
+        self.events: list[models.WorkflowEvent] = []
+
+    def add(self, event: models.WorkflowEvent) -> None:
+        self.events.append(event)
+
+    def get_all_by_tracking_identifier(self, tracking_identifier: str) -> list[models.WorkflowEvent]:
+        return [
+            event for event in self.events
+            if event.tracking_identifier == tracking_identifier
+        ]
+
+
+class WorkflowEvent(Base):
+    __tablename__ = "workflow_event"
+
+    identifier: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True)
+    package_identifier: Mapped[str] = mapped_column(String(30))
+    tracking_identifier: Mapped[str] = mapped_column(String(30))
+    event_type: Mapped[str] = mapped_column(String(30))
+    timestamp: Mapped[datetime] = mapped_column(DateTime(timezone=True))
+    message: Mapped[Optional[str]] = mapped_column(String(1000))
+
+
+class SqlalchemyEventStore(EventStore):
+
+    def __init__(self, session):
+        self.session = session
+
+    def add(self, event: models.WorkflowEvent) -> None:
+        stored_event = WorkflowEvent(
+            identifier=event.identifier,
+            package_identifier=event.package_identifier,
+            tracking_identifier=event.tracking_identifier,
+            event_type=event.event_type.value,
+            timestamp=event.timestamp,
+            message=event.message
+        )
+        self.session.add(stored_event)
+
+    def get_all_by_tracking_identifier(self, tracking_identifier: str) -> list[models.WorkflowEvent]:
+        statement = select(WorkflowEvent).where(
```

Comment:
Should we order by timestamp?

Reply:
That would be reasonable! Was starting to think about sorting and where to do it.

Reply:
And would we want it sorted by chronological/reverse? Would we want to know the last event that was seen?

Reply:
I was thinking of adding another method like `get_latest_for_tracking_id`?

Reply:
LOL I would avoid method explosion and rely on `event_store.get_all_by_tracking_identifier(tracking_identifier)[0]` 🤣

Reply:
Done. https://github.com/mlibrary/dor-py/pull/35/commits/7847f570baf6d6e5844d9b8abb02d1463f162a45